### ISSUE TITLE
Microsoft Graph API Support

### ISSUE THREAD

**Issue by jiasli** [Azure/azure-cli#12946]
[**Azure Active Directory Graph API**](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-graph-api) has been deprecated:

> We strongly recommend that you use [Microsoft Graph](https://developer.microsoft.com/graph) instead of Azure AD Graph API to access Azure Active Directory (Azure AD) resources. Our development efforts are now concentrated on Microsoft Graph and no further enhancements are planned for Azure AD Graph API. There are a very limited number of scenarios for which Azure AD Graph API might still be appropriate; for more information, see the [Microsoft Graph or the Azure AD Graph](https://developer.microsoft.com/office/blogs/microsoft-graph-or-azure-ad-graph/) blog post and [Migrate Azure AD Graph apps to Microsoft Graph](https://docs.microsoft.com/graph/migrate-azure-ad-graph-overview).

We need to switch Azure CLI to use [**Microsoft Graph API**](https://docs.microsoft.com/en-us/graph/api/overview?toc=./ref/toc.json&view=graph-rest-1.0) instead.

**Comment by yonzhan** [comment]
Microsoft Graph API Support

**Comment by jiasli** [comment]
## Call Microsoft Graph API with `az rest`

As [**Microsoft Graph REST API v1.0**](https://docs.microsoft.com/en-us/graph/api/overview?view=graph-rest-1.0) is now GA, we can call it directly with `az rest` to achieve the same effect as `az ad` commands, including all latest features from Microsoft Graph. 

## Examples

### Update `redirectUris` for an [Application](https://docs.microsoft.com/en-us/graph/api/resources/application?view=graph-rest-1.0)

Originally posted at https://github.com/Azure/azure-cli/issues/9501#issuecomment-610979930. We call the [Update application](https://docs.microsoft.com/en-us/graph/api/application-update?view=graph-rest-1.0&tabs=http) API. The GUID part in the following URLs are the object ID of the application. 

```sh
# Get the application
az rest --method GET --uri 'https://graph.microsoft.com/v1.0/applications/b4e4d2ab-e2cb-45d5-a31a-98eb3f364001'

# Update `redirectUris` for `web` property
az rest --method PATCH --uri 'https://graph.microsoft.com/v1.0/applications/b4e4d2ab-e2cb-45d5-a31a-98eb3f364001' --headers 'Content-Type=application/json' --body '{"web":{"redirectUris":["https://myapp.com"]}}'
```

### Add owners to a service principal

Originally posted at https://github.com/Azure/azure-cli/issues/9250#issuecomment-603621148. We call the [servicePrincipal: Add owner](https://docs.microsoft.com/en-us/graph/api/serviceprincipal-post-owners?view=graph-rest-beta&tabs=http) API.

```sh
appId=93dde3da-9fca-47dd-aee2-409b402ffed3
spObjectId=$(az ad sp show --id $appId --query objectId --output tsv)

# Get the object Id for the current user
ownerObjectId=$(az ad signed-in-user show --query objectId -o tsv)

# This applies to both user and service principal as owners
az rest -m POST -u https://graph.microsoft.com/beta/servicePrincipals/$spObjectId/owners/\$ref --headers Content-Type=application/json -b "{\"@odata.id\": \"https://graph.microsoft.com/beta/directoryObjects/$ownerObjectId\"}"

# To add a user as an owner
az rest -m POST -u https://graph.microsoft.com/beta/servicePrincipals/$spObjectId/owners/\$ref --headers Content-Type=application/json -b "{\"@odata.id\": \"https://graph.microsoft.com/beta/users/$ownerObjectId\"}"

# To add a service principal as an owner
az rest -m POST -u https://graph.microsoft.com/beta/servicePrincipals/$spObjectId/owners/\$ref --headers Content-Type=application/json -b "{\"@odata.id\": \"https://graph.microsoft.com/beta/servicePrincipals/$ownerObjectId\"}"
```

### Remarks

- The above commands are for Bash which Azure Cloud Shell uses
- In both PowerShell and CMD terminal, double quotes `"` in the JSON body (as contents) need to be escaped by `\`, even if the JSON body is surrounded by single quotes `'`: `'{\"key\":\"value\"}'`. For more info, please see [Quoting Issues](https://github.com/Azure/azure-cli/blob/dev/doc/use_cli_effectively.md#quoting-issues)

**Comment by phil-holden** [comment]
A comment made by [@jiasli in issue #9250](https://github.com/Azure/azure-cli/issues/9250#issuecomment-603621148) suggests that it should be possible to add a Service Principal as the Owner of an Application using the MS Graph API as follows.

```bash
az rest --method post --uri https://graph.microsoft.com/V1.0/applications/$appObjectId/owners/$ref --headers Content=Type=application/json --body "{\"@odata.id\":\"https://graph.microsoft.com/V1.0/directoryObjects/$spObjectId\"}"
```
this currently returns a Bad Request

```json
{
  "error": {
    "code": "Request_BadRequest",
    "message": "The reference target 'Application_xxx' of type 'Application' is invalid for the 'owners' reference.",
    "innerError": {
      "request-id": "bdf80192-becb-43d5-9447-e0e2a91d7d68",
      "date": "2020-05-27T01:25:14"
    }
  }
}
```
xxx in the response above is the $spObjectId

Is this still a supported scenario using the MS Graph API?

```bash
az ad app owner add --id $appObjectId --owner-object-id $spObjectId
```
returns a similar result.

**Comment by jiasli** [comment]
This error `The reference target 'Application_xxx' of type 'Application' is invalid for the 'owners' reference.` indicates `xxx` is an **application object ID**, not a **service principal object ID**. 

If you are using Azure CLI, retrieve the service principal object ID with

```sh
# Use appId 46eb4122-bd2b-4f54-af7b-6d79b46ee31a
$ az ad sp show --id 46eb4122-bd2b-4f54-af7b-6d79b46ee31a --query objectId --output tsv
8aaa6158-7450-4407-ba08-61377f23d05f
```

If you are using Azure Portal, retrieve the service principal object ID with

![image](https://user-images.githubusercontent.com/4003950/82977586-dfa2cd00-a014-11ea-82bc-3d7002b830b9.png)

![image](https://user-images.githubusercontent.com/4003950/82977673-2690c280-a015-11ea-879a-39fd1d2e2815.png)

Please notice that this issue is only for tracking purpose. For specific issues, you may create a new issue or reply to #9250.

**Comment by phil-holden** [comment]
Thanks @jiasli, works well when you have the right id!

**Comment by Tbohunek** [comment]
@jiasli Thanks man! It solved our problem too! I was ripping my hair for 2 weeks now.

However just to understand.. Before I made the SP1 an owner of SP2, SP1 wasn't able to update the reply urls with error 

`az ad app update --id <appid> --add replyUrls "http://test" --debug`

```
https://graph.windows.net:443 "GET /<tenantid>/applications?$filter=identifierUris%2Fany%28s%3As%20eq%20%<appid>%27%29&api-version=1.6 HTTP/1.1" 200 121
```

`Insufficient privileges to complete the operation.`

This was even despite SP1 having all the permissions on SP2 via [full custom role](https://docs.microsoft.com/en-us/azure/active-directory/users-groups-roles/roles-custom-available-permissions), so I figured it's because it cannot read the Directory itself to list all applications and filter them.

Now when we made the SP1 owner of SP2, it can update the ReplyURLs without having any explicit API permissions nor any custom role as above.

What are we missing? Why does it work without `Directory.Read.All`?

Thank you!

**Comment by jiasli** [comment]
Hi @Tbohunek,

Actually you are hitting a known issue that a `Directory.Read.All` permission should not be mandatory to update an application.

https://github.com/Azure/azure-cli/blob/2b6fe1ee100355dcae6681b01f8101ac81109f8a/src/azure-cli/azure/cli/command_modules/role/custom.py#L1068-L1079

You can see from the code that a tenant query is mandatory before CLI can treat the `--id` as an object ID. 

```
az ───→ identifierUri ───→ objectId   This doesn't work because tenant-level query is needed to search for the app
 │                             ↑
 └─────────────────────────────┘      This works because no tenant-level query is needed
```

This is certainly a **design flaw** and we plan to fix it in the Microsoft Graph integration. We are considering whether we should split `identifierUris`, `appId` and `objectId` to avoid confusions, but this would certainly be a BREAKING CHANGE. 

For example, `az role assignment create` already has this fix. `--assignee-object-id` is used for bypassing Graph permission issues:

```sh
> az role assignment create -h
...
    --assignee                : Represent a user, group, or service principal. supported format:
                                object id, user sign-in name, or service principal name.
    --assignee-object-id      : Use this parameter instead of '--assignee' to bypass graph
                                permission issues. This parameter only works with object ids for
                                users, groups, service principals, and managed identities. For
                                managed identities use the principal id. For service principals, use
                                the object id and not the app id.
```

With that being said, I am not able to repro your scenario. I got `Insufficient privileges to complete the operation.` with both `appId` and `appObjectId` as expected:

```sh
# Owner Service Principal (not Application)
$ ownerObjectId="b0460b1a-d53a-42e8-a6a4-d9f43f0ccc73"

# Owned Application
$ appId="46eb4122-bd2b-4f54-af7b-6d79b46ee31a"
$ appObjectId="9580df2b-3c36-4a15-b909-15bb1b30b3a7"

# Add the owner
$ az rest --method post --uri https://graph.microsoft.com/V1.0/applications/$appObjectId/owners/\$ref --body "{\"@odata.id\":\"https://graph.microsoft.com/V1.0/directoryObjects/$ownerObjectId\"}"

$ az ad app update --id $appId --add replyUrls "http://test"
Insufficient privileges to complete the operation.

$ az ad app update --id $appObjectId --add replyUrls "http://test"
Insufficient privileges to complete the operation.
```

If an app can do tenant-level query without `Directory.Read.All` permission, this is definitely a security breach. 😉

Please check 
1. Whether your are really using the owner application logged in with `az account show`
2. Whether there are remaining permissions left for the owner application

Even as an owner, the owner application still need `Application.ReadWrite.OwnedBy` permission to update its owned applications. For **Microsoft Graph**:

![image](https://user-images.githubusercontent.com/4003950/83100955-bc931e80-a0e3-11ea-8f25-7118caf28378.png)

(Same for AD Graph, but due to the design flaw mentioned above, `az ad app update` won't work.)

To bypass the tenant-level query, you can directly use Microsoft Graph [Update application](https://docs.microsoft.com/en-us/graph/api/application-update?view=graph-rest-1.0&tabs=http) to update the `redirectUris`.

```sh
$ az rest -m PATCH -u https://graph.microsoft.com/v1.0/applications/$appObjectId --body '{"web":{"redirectUris":["https://myapp.com"]}}' --verbose
Request URL: 'https://graph.microsoft.com/v1.0/applications/9580df2b-3c36-4a15-b909-15bb1b30b3a7'
Request method: 'PATCH'
Request headers:
    'User-Agent': 'AZURECLI/2.6.0 (DEB)'
    'Accept-Encoding': 'gzip, deflate'
    'Accept': '*/*'
    'Connection': 'keep-alive'
    'x-ms-client-request-id': '65f37202-02b9-4723-b788-3dda87354bac'
    'Content-Type': 'application/json'
    'CommandName': 'rest'
    'ParameterSetName': '-m -u --body --verbose'
    'Authorization': 'Bearer eyJ0eXAiOiJKV...'
    'Content-Length': '46'
Request body:
{"web":{"redirectUris":["https://myapp.com"]}}
Response status: 204
Response headers:
    'Cache-Control': 'private'
    'Content-Type': 'text/plain'
    'request-id': '180fda8f-739e-41ea-b540-4ea8bfe8bed4'
    'client-request-id': '180fda8f-739e-41ea-b540-4ea8bfe8bed4'
    'x-ms-ags-diagnostic': '{"ServerInfo":{"DataCenter":"Southeast Asia","Slice":"SliceC","Ring":"5","ScaleUnit":"000","RoleInstance":"AGSFE_IN_24"}}'
    'Strict-Transport-Security': 'max-age=31536000'
    'Date': 'Thu, 28 May 2020 05:05:20 GMT'
Response content:

command ran in 1.641 seconds.
```

Again, if you have further questions regarding this specific topic, please kindly create a new issue. I am glad to help. 😊

**Comment by Tbohunek** [comment]
@jiasli Thank you, this helped me find out that I was not using properly isolated CLI sessions, so adding the ReplyURL after making the SP and owner was run by wrong SP.

Now I granted the SP1 the `Azure AD Graph: Application.ReadWrite.OwnedBy` API permission and it works as expected.

Note that in the `azure-cli 2.6.0` command it's visible that it uses the Azure AD Graph and not the new MS Graph:

```
az --version

azure-cli                          2.6.0
command-modules-nspkg              2.0.3
core                               2.6.0
nspkg                              3.0.4
telemetry                          1.0.4


az ad app update --id de9b193b-6b87-4ee8-954e-de1e6e6d99d9  --add replyUrls "http://test" --debug

https://graph.windows.net:443 "GET /<tenantid>/applications?$filter=identifierUris%2Fany%28s%3As%20eq%20%27de9b193b-6b87-4ee8-954e-de1e6e6d99d9%27%29&api-version=1.6 HTTP/1.1" 403 219
````

**Comment by jiasli** [comment]
I am glad it is working for you. 

Yes, `az ad` is still using AD Graph API. This's what this whole issue is talking about. Since Microsoft Graph Service Principal API is GA, we recommend using `az rest` instead of `az ad` for the time being until we fully migrate `az ad` to Microsoft Graph. 

---

However, I tried to add `Microsoft Graph: Application.ReadWrite.OwnedBy` to the owner application, then I am able to list all applications in the tenant using this owner application!

```
> az rest -u 'https://graph.microsoft.com/v1.0/applications' --query "value[].displayName"
[
  "sp19e92232ca5",
  "sp168579505de",
  "sp25f660911937e",
  ...
```

This is against the AAD doc https://docs.microsoft.com/en-us/graph/permissions-reference#remarks-3:

> NOTE: Using the _Application.ReadWrite.OwnedBy_ permission to call `GET /applications` to list applications will fail with a 403.  Instead use `GET servicePrincipals/{id}/ownedObjects` to list the applications owned by the calling application.

**Comment by scals44** [comment]
@jiasli thanks for all the examples of utilizing az rest to call service principals. I'm fighting with a possible edge case now. In an attempt to categorize and programmatically rotate client secrets we're aiming to use Tags on service principal objects. For the most part, this works fine by running az commands like:
`az ad sp update --id appID --add tags "tags:value"`

Before I got to this point I ran into the nanosecond/microsecond Azure Portal bug with az cli and I am getting similar behavior when trying to update tags on a service principal using the above command. The client secrets were set using the "old" method, outlined below:
`                    $servicePrincipal = New-AzureRmADServicePrincipal -ApplicationId $spn.ApplicationId
                    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($servicePrincipal.Secret)
                    $spnPassword = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
                    [Runtime.InteropServices.Marshal]::ZeroFreeBSTR($BSTR)
                    $SecureStringPassword = ConvertTo-SecureString -String $spnPassword -AsPlainText -Force
                    #Apply the password to the app
                    New-AzureRmADAppCredential -ApplicationId $spn.ApplicationId -Password $SecureStringPassword
`

When trying to tag service principals created this way I am finding the generic error: az : Update to existing credential with KeyId 'keyID' is not allowed.

Do you have any working examples of updating tags with the az rest method? I can't seem to get the payload correct. 
`az rest -m "PATCH" -u https://graph.microsoft.com/v1.0/servicePrincipals/objID --headers "Content-Type=application/json" -b '{\"tags\":\""tag:value"\"}'
`

Any help is appreciated, thanks!

**Comment by jiasli** [comment]
@scals44, do you mind creating a new issue for us to track? Thank you.

**Comment by jaslloyd** [comment]
sorry to jump on this thread but is there any easy of updating the redirectUris (by adding one property) using the rest API...it seems like patch will overwrite all of the redirectUrls e.g:

` az rest --method PATCH --uri "https://graph.microsoft.com/v1.0/applications/$OBJECT_ID" --headers 'Content-Type=application/json' --body "{\"spa\":{\"redirectUris\": [\"https://addanotherurlhere.com\"]}}"`

Ends up deleting all the other URLs, since the cli doesn't support updating spa URLs at all yet (only web redirectUris/replyUrls)...

**Comment by rmsy** [comment]
@jaslloyd Someone else may correct me, but I don't believe there is a way to atomically add one URI. However, you can query the Graph API to obtain the existing redirect URIs and use that to formulate your PATCH body:

```bash
az rest --method GET --uri 'https://graph.microsoft.com/v1.0/applications/OBJECT-ID-HERE?$select=spa' --headers 'Content-Type=application/json'
```

returns:

```json
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#applications(spa)/$entity",
  "spa": {
    "redirectUris": [
      "https://myapp.com"
    ]
  }
}
```

With that said, it is disappointing that the CLI itself is so incapable when it comes to many common functions 😕

**Comment by jiasli** [comment]
Hi @jaslloyd, @rmsy's answer is correct. As `spa.redirectUris` is a property, there is no API from MS Graph API to add new entries to it. You need to `GET` the value first, do the addition, then `PATCH` it back.

**Comment by jan-hudec** [comment]
@jaslloyd, here is my build step for ensuring an URI is included:

```bash
SPA_OLD_URIS=$(az rest --method GET --uri "https://graph.microsoft.com/v1.0/applications/$APP_ID" | jq -c '.spa.redirectUris | unique')
SPA_NEW_URIS=$(printf '%s\n' "$SPA_OLD_URIS" | jq -c ". + [\"$NEW_URI\"] | unique")
if [ "x$SPA_NEW_URIS" != "x$SPA_OLD_URIS" ]; then
    SPA_PATCH=$(printf '%s\n' "$SPA_NEW_URIS" | jq -c '{"spa":{"redirectUris":.}}')
    az rest --method PATCH --uri "https://graph.microsoft.com/v1.0/applications/$APP_ID" --headers 'Content-Type=application/json' --body="$SPA_PATCH"
fi
```

(I actually need two URIs for each environment, so I have `jq -c ". + [\"$NEW_URI\", \"$NEW_URI/silent-renew.html\"] | unique"` for the addition).

It could probably be simplified using the built-in JMESPath capability of az to directly calculate the patch in the initial GET, but this was easier to debug.

**Comment by jaslloyd** [comment]
@rmsy @jialindai @jan-hudec thanks for the help, p.s by chance do you happen to know what permissions a service principal needs to update the redirect_uris of itself? I tried adding Application.ReadWrite.OwnedBy but no luck

**Comment by machielvisser** [comment]
Is there an indication when azure cli will be migrated to Microsoft Graph API?

**Comment by Tbohunek** [comment]
Time flies! Still nothing on Microsoft side despite them convincing us to migrate our workloads...
![image](https://user-images.githubusercontent.com/16520646/100857643-36d69e00-348d-11eb-8661-0ae7174d9fd0.png)

**Comment by jiasli** [comment]
Azure CLI team is currently working on the **ADAL -> MSAL migration**. We will start the planning and implementation of **Active Directory Graph -> Microsoft Graph migration** once the previously task is done. + @achandmsft 

Meanwhile, Microsoft Graph team is currently developing their own CLI tool: https://github.com/microsoftgraph/msgraph-cli. This project is still at its early phase. You are very welcome to play with it and share any feedback.

**Comment by Tbohunek** [comment]
@jiasli thanks for the clarification. We are all looking forward. ❤️ 

However, isn't it a bit contra-productive to introduce yet another CLI tool on top of `az`? The confusion between `Get-AzADxxx` and `Get-AzureADxxx` (think `Connect-AzureAD`, `Login-AzAccount`, `Connect-AzAccount`) is already enough. 🤕

**Comment by BasSmuldersCZ** [comment]
We are currently are running into the same issue. Adopting other commands or tools is contra-productive indeed. Please give us a clear timeline.

**Comment by joukevandermaas** [comment]
Pretty crazy that MS makes it hard to use the old thing with new service principals, while there is no way to avoid using the old APIs with automation tooling at the moment. It seems to me that the deprecation is a bit over eager at this point.

If you are only going to support the new graph in March, then why not deprecate the old thing in March? Now all your users have to find a github issue to tell them about a workaround. It's not even in the official documentation.

**Comment by jiasli** [comment]
There is no fully-fledged Python SDK for Microsoft Graph, but only a light-weight client: https://github.com/microsoftgraph/msgraph-sdk-python-core.

During the migration, Azure CLI will use REST call and construct/display the JSON object directly, without building an SDK model.

**Comment by chull434** [comment]
Any update on this? do I update my scripts to use az rest commands or is there actual ms graph commands to use yet?

**Comment by oalimerko** [comment]
is there via `az rest` any replacement for the command `az ad sp credential reset --name <name>`?
With `az ad sp credential reset` i am facing the well known issue `Insufficient privileges to complete the operation.`

**Comment by jiasli** [comment]
@oalimerko, please use `az rest` to invoke Microsoft Graph API like [application: addPassword](https://docs.microsoft.com/en-us/graph/api/application-addpassword?view=graph-rest-1.0&tabs=http) or [servicePrincipal: addPassword](https://docs.microsoft.com/en-us/graph/api/serviceprincipal-addpassword?view=graph-rest-1.0&tabs=http).

**Comment by jiasli** [comment]
## Discussion: whether to preserve deprecate AD Graph parameter names

There has been an internal discussion recently about whether we should **preserve deprecated AD Graph parameter names**.

For Azure PowerShell, after the Microsoft Graph migration, **it uses deprecated AD Graph parameter names as official names, and new Microsoft Graph parameter names as aliases**. For example, according to https://docs.microsoft.com/en-us/powershell/module/az.resources/new-azadapplication?view=azps-7.2.0, `New-AzADApplication` uses the deprecated AD Graph property name `-HomePage` as the official parameter name:

![image](https://user-images.githubusercontent.com/4003950/156689775-9744a332-df6d-4230-8b52-02e36841d275.png)

According to https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-property-differences, `homepage` has been replaced by `web/homePageUrl` in Microsoft Graph:

![image](https://user-images.githubusercontent.com/4003950/155959373-2e1e8b65-c2e7-48ac-a681-3f442352f944.png)

The new corresponding parameter name should be `-WebHomePageUrl`, but `-WebHomePageUrl` is used as an alias, instead of the official name. Other parameters face similar problems:

- `-AvailableToOtherTenants` which conflicts with `-SignInAudience`
- `-ReplyUrls` vs `-WebRedirectUri`. `New-AzADApplication` makes `-ReplyUrls` and `-WebRedirectUri` the same:
  ![image](https://user-images.githubusercontent.com/4003950/155960002-0b35e72c-8db0-452f-bfbd-0f07b982a672.png)
  However, in Microsoft Graph, `replyUrls` is split into `web/redirectUris` and `publicClient/redirectUris`, instead of replaced by `web/redirectUris`: ![image](https://user-images.githubusercontent.com/4003950/155959675-53781296-4b22-4606-ae0c-4d3cd19042a0.png)

### Problems

- New users of Microsoft Graph can't easily find the corresponding parameter names (`-WebHomePageUrl`). Instead, they will have to learn the deprecated AD Graph parameter names (`-HomePage`). This increases learning effort for new users.
- The cmdlet parameter doesn't match the output - `-HomePage` maps to `web/homePageUrl`, introducing inconsistency.

### Azure CLI's solution: be consistent with new Microsoft Graph API

Currently, as detailed in https://docs.microsoft.com/cli/azure/microsoft-graph-migration, we decide to replace deprecated AD Graph parameter names with new Microsoft Graph ones, such as

> `--homepage` argument is replaced by `--web-home-page-url`

If you have any feedback, please feel free to left a comment in this issue.

**Comment by Tbohunek** [comment]
@jiasli Would it be sensible to keep the old names as aliases where it makes sense? Like I'm not looking forward to typing `--web-home-page-url` not to mention I'm not going to remember it.

**Comment by jiasli** [comment]
For `--homepage`, we can surely do it, as it is a one-one mapping to `--web-home-page-url`, unlike `--reply-urls` which doesn't have a definitive mapping:

> `--reply-urls` argument is split into `--web-redirect-uris` and `--public-client-redirect-uris`

@maartengo, @anarsen, @bbkane, any thoughts?

**Comment by Tbohunek** [comment]
The idea was that alias follows the current/original use, so  `--reply-urls` would alias to `--web-redirect-uris`, but I see this isn't very popular idea, so never mind.

**Comment by jiasli** [comment]
> but I see this isn't very popular idea, so never mind.

No, no. This is definitely a very constructive idea. That's exactly why I am so hesitating on the change. 😊

**Comment by jiasli** [comment]
Thanks @joukevandermaas for the feedback.

> The AD graph change is already producing huge churn in our organization for no perceivable benefit on our side.

I totally understand your concern. This is very valuable user voice. As you can see from this GitHub issue, this is the decision of AAD Graph service and we Azure CLI team also have been spending lots of effort adopting the new Microsoft Graph API, but I believe Microsoft Graph is introducing more functionalities and will definitely be a better successor to AD Graph: https://docs.microsoft.com/en-us/graph/overview

That said, since the whole `az ad` command module will undergo larging amount of breaking changes, keeping old AD Graph names will do little benefit and will even introduce input-output inconsistency.

**Comment by lkurzyniec** [comment]
Please support also `Get-MgContext` command.

https://docs.microsoft.com/en-us/graph/powershell/app-only?tabs=azure-portal#authenticate
> Run `Get-MgContext` to verify that you've authenticated with app-only. The output should look like the following.
> 
> ```PowerShell
> ClientId              : YOUR_APP_ID
> TenantId              : YOUR_TENANT_ID
> CertificateThumbprint :
> Scopes                : {Group.Read.All, User.Read.All}
> AuthType              : AppOnly
> CertificateName       : YOUR_CERT_SUBJECT
> Account               :
> AppName               : Graph PowerShell Script
> ContextScope          : Process
> ```

**Comment by jiasli** [comment]
@lkurzyniec, this doesn't seem to to related to Microsoft Graph migration. Similar Azure CLI command is `az account show`.

**Comment by lkurzyniec** [comment]
@jiasli got it, you are right

**Comment by pilarragorri-portugal** [comment]
> Thanks @joukevandermaas for the feedback.
> 
> > The AD graph change is already producing huge churn in our organization for no perceivable benefit on our side.
> 
> I totally understand your concern. This is very valuable user voice. As you can see from this GitHub issue, this is the decision of AAD Graph service and we Azure CLI team also have been spending lots of effort adopting the new Microsoft Graph API, but I believe Microsoft Graph is introducing more functionalities and will definitely be a better successor to AD Graph: https://docs.microsoft.com/en-us/graph/overview
> 
> That said, since the whole `az ad` command module will undergo larging amount of breaking changes, keeping old AD Graph names will do little benefit and will even introduce input-output inconsistency.

Hey  @jiasli, I completely agree with the fact that Graph is going to be a better successor to AD Graph.
My main concern here is the lack of clarity on what is actually going to change in regards of the 'az ad' module behaviours, do we have some sort of list with the expected changes?

**Comment by jiasli** [comment]
> My main concern here is the lack of clarity on what is actually going to change in regards of the 'az ad' module behaviours, do we have some sort of list with the expected changes?

Surely we have! 😊 Please see https://docs.microsoft.com/en-us/cli/azure/microsoft-graph-migration

**Comment by RamonaStefan** [comment]
Hey @jiasli! In this [documentation](https://docs.microsoft.com/en-us/cli/azure/microsoft-graph-migration), there are no breaking changes mentioned on the `az ad sp show` command. However, the warning appears for the entire `az` module as seen in this [PR](https://github.com/Azure/azure-cli/pull/21302/files#diff-577bf256565532649f8412372d893504965ad72b24513b33f24e0aa8d65dc2b1). This is blocking our scripts. Could you please fix it? Or is there any replacement for the command?

**Comment by jiasli** [comment]
@RamonaStefan, your script shouldn't decide whether a command is successful based on `stderr`. It should check the exit code instead. Please see https://github.com/Azure/azure-cli#exit-codes

If you want to disable the warning, there is a global argument `--only-show-errors` to disable warnings for a specific command. You may also disable warnings for all commands with

```
az config set core.only_show_errors=True
```

For more info, please see https://docs.microsoft.com/en-us/cli/azure/azure-cli-configuration

**Comment by acefei** [comment]
@jiasli I tried to use `az rest --method POST --uri https://graph.microsoft.com/v1.0/applications/AppObjectId/addPassword --body '{"passwordCredential":{"displayName": "xxxx"}}'` with the required permission.
![image](https://user-images.githubusercontent.com/1448772/163377013-9e99e3d8-16f8-4241-8561-a80a72b183d8.png)
It still shows `Insufficient privileges to complete the operation.`
Any thoughts?

Reproduce steps:
1. az login --service-principal -u <Application (client) ID> -p xxx --tenant xxx
2. az rest --method POST --uri https://graph.microsoft.com/v1.0/applications/AppObjectId/addPassword --body '{"passwordCredential":{"displayName": "xxxx"}}'    (shows  `Insufficient privileges to complete the operation.` )
3. az rest -m POST -u https://graph.microsoft.com/v1.0/applications/AppObjectId/owners/\$ref -b "{\"@odata.id\": \"https://graph.microsoft.com/v1.0/directoryObjects/AppObjectId\"}" (shows  `Insufficient privileges to complete the operation.` )

**Comment by jiasli** [comment]
@acefei, Azure CLI is merely a REST client. For permission issue, you may reach out to AAD support: https://azure.microsoft.com/en-us/support/create-ticket/

**Comment by drdamour** [comment]
@acefei what shell you running that from?

**Comment by Tbohunek** [comment]
For me it works. Just running from PowerShell.
@acefei did you grant/obtain admin consent on `Application.ReadWrite.OwnedBy`? That is required.
You can remove `Directory.Read.All Delegated`, that isn't used here.

There's however another problem, https://github.com/Azure/azure-cli/issues/9742, so you need to call `--body '{\"passwordCredential\":{\"displayName\": \"xxxx\"}}'` instead. 😞

**Comment by Tbohunek** [comment]
@jiasli There's one more problem..
`az rest --method POST --uri https://graph.microsoft.com/v1.0/applications/<app_objectid>/addPassword --body '{\"passwordCredential\":{\"displayName\": \"mytest6\"}}'`

> Bad Request({"error":{"code":"TooManyAppPasswords","message":"Not allowed to add 3 passwords. You can only add a maximum of 2 passwords.","details":[{"target":"passwordCredentials","code":"TooManyAppPasswords"}],"innerError":{"date":"2022-04-15T17:18:50","request-id":"f0c9aaac-9875-403e-bbd9-9048ec006efb","client-request-id":"f0c9aaac-9875-403e-bbd9-9048ec006efb"}}})

Same error also when adding Secret via Portal.

It seems this only happens for Apps created with `az rest --method POST --uri https://graph.microsoft.com/v1.0/applications --body '{\"displayName\":\"myclicreated\"}'`.
When I create App via Portal and I make the SP its owner with `az ad app owner add`, then I can use the above `az rest` or Portal to add as many passwords as I please.

Any clue what's behind that? Is this the right repo for this issue?

**Comment by jiasli** [comment]
@Tbohunek, I believe the 3 passwords limitation is a design of Graph API [application: addPassword](https://docs.microsoft.com/en-us/graph/api/application-addpassword?view=graph-rest-1.0&tabs=http).

I can reproduce the same. Let me internally involve Graph team on this issue.

**Comment by Tbohunek** [comment]
@jiasli I don't think so. The Graph API `addPassword` works on a App **created** via Azure Portal. It has to do with the App config but comparing its json with Graph-API created App doesn't reveal any clues.

**Comment by Tbohunek** [comment]
I wonder what's your problem @joukevandermaas 👎

**Comment by joukevandermaas** [comment]
@Tbohunek you are spamming an issue that many people are monitoring with something only tangentially related. Can you not imagine why I might find that annoying? Why not take the unrelated issues somewhere else so those of use who care about the original topic can get notified when there are actual updates on this issue?

I tried to indicate this in a way that does not update everyone who follows this issue but I guess that didn't work.

**Comment by acefei** [comment]
> @acefei what shell you running that from? 
I ran `az rest` in `mcrmcr.microsoft.com/azure-cli` in container.

**Comment by acefei** [comment]
> @acefei, Azure CLI is merely a REST client. For permission issue, you may reach out to AAD support: https://azure.microsoft.com/en-us/support/create-ticket/

Unfortunately, I did that, but I can't find AAD related in the problem type with `Technical` issue type.

BTW, I don't think it was a typical permission issue. As we know, there are two type applications with the same Client ID
once the SPN created, Enterprise applications (service principals) and App registrations (application).
so they are using different [Microsoft Graph API](https://docs.microsoft.com/en-us/graph/overview?view=graph-rest-1.0).

In my problem, with the same client ID and same permissions that following the requirement in API page, the permission issue only occurs on [`POST /applications/{id}/owners/$ref`](https://docs.microsoft.com/en-us/graph/api/application-post-owners?view=graph-rest-1.0&tabs=http), but passed on [`POST /servicePrincipals/{id}/owners/$ref`](https://docs.microsoft.com/en-us/graph/api/serviceprincipal-post-owners?view=graph-rest-1.0&tabs=http)

**Comment by acefei** [comment]
> For me it works. Just running from PowerShell with Az CLI installed. @acefei did you grant/obtain admin consent on `Application.ReadWrite.OwnedBy`? That is required. You can remove `Directory.Read.All Delegated`, that isn't used here.
> 
> There's however another problem, #9742, so you need to call `--body '{\"passwordCredential\":{\"displayName\": \"xxxx\"}}'` instead. 😞

I think my original thread was clear.
1. I've obtain admin consent on `Application.ReadWrite.OwnedBy` (see the screenshot)
2. Do you mentioned the escape characters? yes I did, but the problem persists.
Literally, from the `Application.ReadWrite.OwnedBy` permission, the operator should be the owner of the application, so I tried to add appId into its owner list by `az rest -m POST -u https://graph.microsoft.com/v1.0/applications/$AppObjectId/owners/\$ref -b "{\"@odata.id\": \"https://graph.microsoft.com/v1.0/directoryObjects/$AppObjectId\"}"`, but that return `Insufficient privileges to complete the operation`

**Comment by Tbohunek** [comment]
@acefei your screenshot does not show whether the consent was actually **granted**, just that permission is configured. Add a wider screenshot.
Instead of modifying an existing App, try creating new App. Does that not work either?

**Comment by ABerTSC** [comment]
is there any update now i still get "Insufficient privileges to complete the operation" trying to use

`# Starter pipeline
# Start with a minimal pipeline that you can customize to build and deploy your code.
# Add steps that build, run tests, deploy, and more:
# https://aka.ms/yaml

trigger:
- main

pool:
  vmImage: ubuntu-latest

steps:
- task: AzureCLI@2
  inputs:
    azureSubscription: SC-ARM-001
    scriptType: pscore
    scriptLocation: inlineScript
    inlineScript: |
      # Place scripting here
      $appId = (az ad app create --display-name AuthTest --homepage https://mhauthtest.azurewebsites.net --identifier-uris https://mhauthtest.azurewebsites.net --reply-urls https://mhauthtest.azurewebsites.net/.auth/login/aad/callback --query 'appId' -o tsv)
      Write-Host "App ID: $appId"`
at Azure DevOps Pipeline

**Comment by munntjlx** [comment]
I found the 'updated' version for this: https://learn.microsoft.com/en-us/graph/api/rbacapplication-post-roleassignmentschedulerequests?view=graph-rest-1.0&tabs=powershell

**Comment by dsantunes** [comment]
> @jiasli There's one more problem.. `az rest --method POST --uri https://graph.microsoft.com/v1.0/applications/<app_objectid>/addPassword --body '{\"passwordCredential\":{\"displayName\": \"mytest6\"}}'`
> 
> > Bad Request({"error":{"code":"TooManyAppPasswords","message":"Not allowed to add 3 passwords. You can only add a maximum of 2 passwords.","details":[{"target":"passwordCredentials","code":"TooManyAppPasswords"}],"innerError":{"date":"2022-04-15T17:18:50","request-id":"f0c9aaac-9875-403e-bbd9-9048ec006efb","client-request-id":"f0c9aaac-9875-403e-bbd9-9048ec006efb"}}})
> 
> Same error also when adding Secret via Portal.
> 
> It seems this only happens for Apps created with `az rest --method POST --uri https://graph.microsoft.com/v1.0/applications --body '{\"displayName\":\"myclicreated\"}'`. When I create App via Portal and I make the SP its owner with `az ad app owner add`, then I can use the above `az rest` or Portal to add as many passwords as I please.
> 
> Any clue what's behind that? Is this the right repo for this issue?

Did you ever find the cause for this?
